app/services/proforma_service/convert_task_to_exercise.rb

```diff
@@ -60,44 +60,56 @@ def string_to_bool(str)
-      model_solution_files + test_files + task_files.values.tap {|array| array.each {|file| file.role ||= 'regular_file' } }
+      model_solution_files + test_files + task_files
     end
 
     def test_files
-      @task.tests.map do |test_object|
-        task_files.delete(test_object.files.first.id).tap do |file|
-          file.weight = extract_meta_data(test_object.meta_data&.dig('test-meta-data'), 'weight').presence || 1.0
-          file.feedback_message = extract_meta_data(test_object.meta_data&.dig('test-meta-data'), 'feedback-message').presence || 'Feedback'
-          file.hidden_feedback = extract_meta_data(test_object.meta_data&.dig('test-meta-data'), 'hidden-feedback').presence || false
-          file.role ||= 'teacher_defined_test'
+      @task.tests.flat_map do |test|
+        test.files.map do |task_file|
+          codeocean_file_from_task_file(task_file, test).tap do |file|
+            file.weight = extract_meta_data(test.meta_data&.dig('test-meta-data'), 'test-file', task_file.id, 'weight').presence || 1.0
+            file.feedback_message = extract_meta_data(test.meta_data&.dig('test-meta-data'), 'test-file', task_file.id, 'feedback-message').presence || 'Feedback'
+            file.hidden_feedback = extract_meta_data(test.meta_data&.dig('test-meta-data'), 'test-file', task_file.id, 'hidden-feedback').presence || false
+            file.role = 'teacher_defined_test' unless file.teacher_defined_assessment?
+          end
         end
       end
     end
 
     def model_solution_files
-      @task.model_solutions.map do |model_solution_object|
-        task_files.delete(model_solution_object.files.first.id).tap do |file|
-          file.role ||= 'reference_implementation'
+      @task.model_solutions.flat_map do |model_solution|
+        model_solution.files.map do |task_file|
+          codeocean_file_from_task_file(task_file, model_solution).tap do |file|
+            file.role ||= 'reference_implementation'
+          end
         end
       end
     end
 
     def task_files
-      @task_files ||= @task.all_files.reject {|file| file.id == 'ms-placeholder-file' }.to_h do |task_file|
-        [task_file.id, codeocean_file_from_task_file(task_file)]
+      @task.files.reject {|file| file.id == 'ms-placeholder-file' }.map do |task_file|
+        codeocean_file_from_task_file(task_file).tap do |file|
+          file.role ||= 'regular_file'
+        end
       end
+      # @task_files ||= @task.all_files.reject {|file| file.id == 'ms-placeholder-file' }.to_h do |task_file|
+      #   [task_file.id, codeocean_file_from_task_file(task_file)]
+      # end
     end
 
-    def codeocean_file_from_task_file(file)
+    def codeocean_file_from_task_file(file, parent_object = nil)
       extension = File.extname(file.filename)
-      codeocean_file = CodeOcean::File.new(
+
+      codeocean_file = CodeOcean::File.where(context: @exercise).where('xml_id_path LIKE ?', "%#{file.id}").first_or_initialize(context: @exercise)
```

Comment:
The `LIKE` query looks dangerous. If I have a `file.id` of `1` (which is possible, since CodeOcean uses increasing numbers starting from 1 actually), any `xml_id_path` ending in `1` belonging to the current exercise is found. 

Otherwise, you could also try to use `@exercise.where(...)` to make it a little shorter.

Reply:
I made the query more specific, but am not sure how to use `@exercise.where` here, since I'm looking for files

Reply:
Thanks, the query looks better now. Can you add a comment to to the `where`? I was just figuring it out myself, but the syntax is not very common, I'd say:

```rb
# Find file by the given XML ID. Since the `xml_id_path` might contain multiple values (for the containing ProFormA test or model solution), the actual XML ID of the file is always the last entry in the `xml_id_path`  
```

> but am not sure how to use @exercise.where here, since I'm looking for files

Sorry, I missed one important part. The following should work: `@exercise.files.where('array_length(xml_id_path, 1) IS NOT NULL AND xml_id_path[array_length(xml_id_path, 1)] = ?', file.id).first_or_initialize(context: @exercise)`